### PR TITLE
Fixup/oracle dates

### PR DESCRIPTION
**Pull request by detule**
Fixes long-standing ORACLE issues with using batch sizes greater than one when writing to tables with DATE and TIMESTAMP fields.

See, for example #349, #350, #391

### CODE REVIEW

File: R/driver-oracle.R [hunk 1]
```diff
@@ -122,7 +122,12 @@ setMethod("odbcConnectionColumns_", c("Oracle", "character"),
       )
     }
 
-    dbGetQuery(conn, query)
+    res <- dbGetQuery(conn, query)
+
+    res$data_type <- as.numeric(res$data_type)
+    res[res$`field.type` == "DATE", c("data_type", "column_size")] <- c(91, 6)
+    res[grepl("TIMESTAMP", res$`field.type`), c("data_type", "column_size")] <- c(93, 16)
```
Comment (simonpcouch) [inline R/driver-oracle.R:129]:
```suggestion
    res[res$field.type == "DATE", c("data_type", "column_size")] <- c(91, 6)
    res[grepl("TIMESTAMP", res$field.type), c("data_type", "column_size")] <- c(93, 16)
```

No need for backticks with a `.` :)

``` r
x <- data.frame(dotted.column = 1)
x$dotted.column
#> [1] 1
```

<sup>Created on 2024-05-28 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: tests/testthat/test-driver-oracle.R [hunk 2]
```diff
@@ -18,3 +19,19 @@ test_that("can detect existence of table", {
...
+  # See #349, #350, #391
+  con <- test_con("ORACLE")
+
+  ir <- datasets::iris
+  values <- data.frame(
+    datetime = as.POSIXct(as.numeric(ir$Petal.Length * 10), origin = "2024-01-01", tz = "UTC"),
+    date    = as.Date(as.numeric(ir$Petal.Length * 10), origin = "2024-01-01", tz = "UTC"),
+    integer = as.integer(ir$Petal.Width * 100),
+    double = ir$Sepal.Length,
+    varchar = ir$Species)
```
Comment (simonpcouch) [inline tests/testthat/test-driver-oracle.R:33]:
```suggestion
  values <- data.frame(
    datetime = as.POSIXct(as.numeric(iris$Petal.Length * 10), origin = "2024-01-01", tz = "UTC"),
    date    = as.Date(as.numeric(iris$Petal.Length * 10), origin = "2024-01-01", tz = "UTC"),
    integer = as.integer(iris$Petal.Width * 100),
    double = iris$Sepal.Length,
    varchar = iris$Species)
```

Objects from datasets are available without namespacing!

File: R/driver-oracle.R (hunk 1 above)
Comment (simonpcouch) [reply R/driver-oracle.R:129]: Could we also drop an explanatory comment above these? e.g.

```suggestion
    # Overwrite type and precision/size using ODBC standard codes (#349, #350, #391).
    res[res$field.type == "DATE", c("data_type", "column_size")] <- c(91, 6)
    res[grepl("TIMESTAMP", res$field.type), c("data_type", "column_size")] <- c(93, 16)
```

File: R/driver-oracle.R [hunk 3]
```diff
@@ -122,7 +122,12 @@ setMethod("odbcConnectionColumns_", c("Oracle", "character"),
       )
     }
 
-    dbGetQuery(conn, query)
+    res <- dbGetQuery(conn, query)
+
+    res$data_type <- as.numeric(res$data_type)
```
Comment (simonpcouch) [inline R/driver-oracle.R:127]: Do we need to revert this column back to another type after processing it?

**Review by simonpcouch — APPROVED** [review]
Unable to push this around locally before the upcoming release, but these changes feel like a step forward for me!

File: R/driver-oracle.R (hunk 1 above)
Comment (hadley) [reply R/driver-oracle.R:129]: And maybe a link to where the field type numbers come from?

Comment (detule) [reply R/driver-oracle.R]: Done, thanks

File: R/driver-oracle.R (hunk 3 above)
Comment (detule) [reply R/driver-oracle.R:135]: Hey Simon:

Good question - I think the query can/does return all `NULL`s for that field, which in turn I think gives a column full of `NA_character_` in the data-frame.

File: R/driver-oracle.R [hunk 4]
```diff
@@ -122,7 +122,20 @@ setMethod("odbcConnectionColumns_", c("Oracle", "character"),
...
+    # fields ( nor does our custom query above ).  We use these data
+    # as a fallback for SQLDescribeParam ( also incorrectly describes
+    # parameters for ORACLE ) when writing using prepared statements ( see
+    # `dbAppendTable` ).  Fix them here.  91 and 93 are SQL_TYPE_DATE,
+    # and SQL_TYPE_TIMESTAMP, respectively in sql.h.  Six and sixteen
+    # are the sizes of nanodbc::date and nanodbc::timestmp in bytes
+    # (#349, #350, #391).
+    res$data_type <- as.numeric(res$data_type)
+    res[res$field.type == "DATE", c("data_type", "column_size")] <- c(91, 6)
+    res[grepl("TIMESTAMP", res$field.type), c("data_type", "column_size")] <- c(93, 16)
```
Comment (hadley) [inline R/driver-oracle.R:137]: This is a very minor quibble but I don't trust this sort of cross-column subset-assignment, so I'd prefer to see something more like:

```R
res$data_type[res$field.type == "DATE"] <- 91
res$column_size[res$field.type == "DATE"] <- 6
```

I suspect that might also eliminate the need for `as.numeric(res$data_type)` above.

**Review by hadley — APPROVED** [review]

File: R/driver-oracle.R (hunk 4 above)
Comment (detule) [reply R/driver-oracle.R]: Thanks.  Done - let me know if that's not what you had in mind.

I think the `as.numeric` may still be needed, but happy to update if you guys think there's a better way.

```
 ┃  Browse[1]> res <- dbGetQuery(conn, query)
 ┃  Browse[1]> str(res)
 ┃  'data.frame':   5 obs. of  17 variables:
 ┃   $ name                   : chr  "datetime" "date" "integer" "double" ...
 ┃   $ field.type             : chr  "TIMESTAMP(6)" "DATE" "NUMBER" "BINARY_DOUBLE" ...
 ┃   $ table_name             : chr  "test_table" "test_table" "test_table" "test_table" ...
 ┃   $ schema_name            : chr  "SA" "SA" "SA" "SA" ...
 ┃   $ catalog_name           : chr  NA NA NA NA ...
 ┃   $ data_type              : chr  NA NA NA NA ...
 ┃   $ column_size            : num  NA NA NA NA 255
 ┃   $ buffer_length          : num  11 16 40 8 255
 ┃   $ decimal_digits         : num  6 NA 0 NA NA
 ┃   $ numeric_precision_radix: chr  NA NA NA NA ...
 ┃   $ remarks                : chr  NA NA NA NA ...
 ┃   $ column_default         : chr  NA NA NA NA ...
 ┃   $ sql_data_type          : chr  NA NA NA NA ...
 ┃   $ sql_datetime_subtype   : chr  NA NA NA NA ...
 ┃   $ char_octet_length      : num  0 0 0 0 255
 ┃   $ ordinal_position       : num  1 2 3 4 5
 ┃   $ nullable               : num  1 1 1 1 1
 ┃  Browse[1]>     isDate <- res$field.type == "DATE"
 ┃      res$data_type[isDate] <- 91
 ┃      res$column_size[isDate] <- 6
 ┃      isTimestamp <- grepl("TIMESTAMP", res$field.type)
 ┃      res$data_type[isTimestamp] <- 93
 ┃      res$column_size[isTimestamp] <- 16
 ┃  Browse[1]> str(res)
 ┃  'data.frame':   5 obs. of  17 variables:
 ┃   $ name                   : chr  "datetime" "date" "integer" "double" ...
 ┃   $ field.type             : chr  "TIMESTAMP(6)" "DATE" "NUMBER" "BINARY_DOUBLE" ...
 ┃   $ table_name             : chr  "test_table" "test_table" "test_table" "test_table" ...
 ┃   $ schema_name            : chr  "SA" "SA" "SA" "SA" ...
 ┃   $ catalog_name           : chr  NA NA NA NA ...
 ┃   $ data_type              : chr  "93" "91" NA NA ...
 ┃   $ column_size            : num  16 6 NA NA 255
 ┃   $ buffer_length          : num  11 16 40 8 255
 ┃   $ decimal_digits         : num  6 NA 0 NA NA
 ┃   $ numeric_precision_radix: chr  NA NA NA NA ...
 ┃   $ remarks                : chr  NA NA NA NA ...
 ┃   $ column_default         : chr  NA NA NA NA ...
 ┃   $ sql_data_type          : chr  NA NA NA NA ...
 ┃   $ sql_datetime_subtype   : chr  NA NA NA NA ...
 ┃   $ char_octet_length      : num  0 0 0 0 255
 ┃   $ ordinal_position       : num  1 2 3 4 5
 ┃   $ nullable               : num  1 1 1 1 1
```
This is a bit of a function of the silly way the `data_type` column is formulated in the query ( always `NULL` ) - we can probably do something better to make sure that the column comes back as numeric, but there is also probably some value in keeping the query as close as the original `SQLColumns` implementation for the OEM oracle driver.

**Pull request merged**